### PR TITLE
YAML Editor enhancements

### LINKED ISSUE #36: Control Plane Tweaking
# Story
- As user I want to tweak certain features like alpha/beta switches or allowing privileged containers or not, so that I can do more things with my clusters, even though not all are production-ready yet in Kubernetes.

# Motivation
The Gardener allows to tweak the configurations of API server, controller manager, scheduler and kube-proxy as well as allowing privileged containers or not, but we haven't exposed that yet (unless we are fine with #34).

# Acceptance Criteria
To be defined in more detail. We need to investigate first (with Rafael F.), whether we can offer some pre-selection or how to actually allow the configuration (outcome can also be to leave it at #34, but then we need some documentation and it will not be very convenient).
- [ ] During cluster creation/update the user shall be able to switch on/off privileged containers support
- [ ] ...

# Resources
See [Gardener](https://github.wdf.sap.corp/kubernetes-attic/garden-operator/issues/198).

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual test tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests or added manual test tickets?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the operations guide about ops-relevant changes?
- [ ] User documentation: Have you informed the documentation developers about user-relevant changes?

### LINKED ISSUE #419: Specify OIDC configuration on Shoot creation/update
At this moment it's impossible (via the creation UI) to specify the OpenId Connect configuration of the `kube-apiserver`.
Specifying OIDC is key requirement in removing the need for all members of an account to be `cluster-admins`.
Therefore the I would recoomend the following 3 options to be added in the dashboard and visible when the cluster creation / edit dialog is active:

## Specify configuration

In this section of the dialog, the following options should be available:

- **Parameter** `Issuer` 

  **Description** URL of the provider which allows the API server to discover public signing keys. Only URLs which use the `https://` scheme are accepted.  This is typically the provider's discovery URL without a path, for example "https://dex.example.com".  This URL should point to the level below .well-known/openid-configuration 

  **Example** If the discovery URL is `https://dex.example.com/.well-known/openid-configuration`, the value should be `https://dex.example.com` 

  **Required** Yes 

- **Parameter** `Client ID` 

  **Description** A client id that all tokens must be issued for. 

  **Example** `kubernetes` 

  **Required** Yes 

- **Parameter** `Username Claim` 

  **Description** JWT claim to use as the user name. By default `sub`, which is expected to be a unique identifier of the end user. Admins can choose other claims, such as `email` or `name`, depending on their provider. However, claims other than `email` will be prefixed with the issuer URL to prevent naming clashes with other plugins. 

  **Example** `sub` 

  **Required** No 

- **Parameter** `Username Prefix` 

  **Description** Prefix prepended to username claims to prevent clashes with existing names (such as `system:` users). For example, the value `oidc:` will create usernames like `oidc:jane.doe`. If this flag isn't provided and `User Claim` is a value other than `email` the prefix defaults to `( Issuer URL )#` where `( Issuer URL )` is the value of `Issuer`. The value `-` can be used to disable all prefixing. 

  **Example** `oidc:` 

  **Required** No 

- **Parameter** `Groups Claim` 

  **Description** JWT claim to use as the user's group. If the claim is present it must be an array of strings. 

  **Example** `groups`

  **Required** No 

- **Parameter** `Groups Prefix` 

  **Description** Prefix prepended to group claims to prevent clashes with existing names (such as `system:` groups). For example, the value `oidc:` will create group names like `oidc:engineering` and `oidc:infra`. 

  **Example** `oidc:` 

  **Required**  No 

- **Parameter** `Required claim` 

  **Description** A key=value pair that describes a required claim in the ID Token. If set, the claim is verified to be present in the ID Token with a matching value. Repeat this flag to specify multiple claims. 

  **Example** `claim=value` 

  **Required** No 

- **Parameter** `CA Bundle` 

  **Description** Certificate for the CA that signed your identity provider's web certificate.  Defaults to the host's root CAs. 

  **Example** 

  ```
  -----BEGIN CERTIFICATE-----
  MIIBizCCATGgAwIBAgIUH4plk9qwD61FVXgiOTngFU5FeSkwCgYIKoZIzj0EAwIw
  EjEQMA4GA1UEAxMHUm9vdC1DQTAgFw0xNjEwMTEwNTA2MDBaGA8yMTE2MDkxNzA1
  MDYwMFowEjEQMA4GA1UEAxMHUm9vdC1DQTBZMBMGByqGSM49AgEGCCqGSM49AwEH
  A0IABI2CsrAnMGT8P2VGU2MLo5pv86Z74kcV9hgkLJUkSaeNyc1s89w7X5V2wvwu
  iWEJRGm5RoZJausmyZLZEoKEVXejYzBhMA4GA1UdDwEB/wQEAwIBBjAPBgNVHRMB
  Af8EBTADAQH/MB0GA1UdDgQWBBQez+vouP6XzY6mYCb/23YAOxXuLTAfBgNVHSME
  GDAWgBQez+vouP6XzY6mYCb/23YAOxXuLTAKBggqhkjOPQQDAgNIADBFAiBGclts
  vJRM+QMVoV/1L9b+hvhgLIp/OupUFsSOReefIwIhALY06hBklyh8eFwuBtyX2VcE
  8xlVn4/5idUvc3Xv2h9s
  -----END CERTIFICATE-----
  ```

  **Required** No 

- **Parameter** `Signing Algorithms` 

  **Description** Comma-separated list of allowed JOSE asymmetric signing algorithms. JWTs with a 'alg' header value not in this list will be rejected. Values are defined by RFC 7518 https://tools.ietf.org/html/rfc7518#section-3.1. Defaults to `RS256` 

  **Example** `RS256` 

  **Required**  No 

## Inherit Gardener API Server configuration.

This uses the OIDC configuration of the  Gardener Kubernetes API Server. It **MUST** be provided as configuration to the dashboard on installation, otherwise this option should not be displayed or be grayed-out.

There shoud be a warning indicating that if this is chosen, **ALL** users which can be be authenticated to dashboard can be authenticated to the respective Shoot cluster and it's up to the Project Admins to configure `RBAC`.

> Note: I really would like this option to be default (configurable) option. It can improve security a lot.

## Do not configure.

Does nothing. Current behavior.

There shoud be a warning indicating that if this is chosen, the Garden dashboard cannot provide personalized `kubeconfig` to Project Members.

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Introduces Code Completion and Tooltips in Shoot Editor.

**Which issue(s) this PR fixes**:
Fixes #36 Fixes #419

**Special notes for your reviewer**:
Screen Shot 2019-09-25 at 12 22 34
Screen Shot 2019-09-25 at 12 22 27
Screen Shot 2019-09-25 at 12 22 06

**Release note**:

```improvement user
The dashboard now supports code completion and shows information about keywords in the cluster YAML editor
```

### CODE REVIEW

File: LICENSE.md [hunk 1]
```diff
@@ -341,6 +341,10 @@ Copyright (c) 2011 Einar Otto Stangvik <einaros@gmail.com>\
 Copyright (c) 2014 Code Charm\
 [MIT License](https://github.com/encharm/x509.js/blob/master/LICENSE)
 
+- [swagger-parser](https://github.com/APIDevTools/swagger-parser)\
```
Comment (petersutter) [inline LICENSE.md]: move entry under socket.io

File: backend/lib/routes/info.js [hunk 2]
```diff
@@ -58,3 +59,29 @@ async function fetchGardenerVersion () {
...
+
+let shootSpec // Cache, TODO: Need to update cache when apiserver gets updated
+
+async function fetchShootSpec (user) {
+  if (shootSpec) {
+    return shootSpec
+  }
+  const client = user.client
+
+  const res = await client.openAPI.get()
```
Comment (petersutter) [inline backend/lib/routes/info.js]: I would not use user's client to fetch the openAPI as we use the result also for other users and instead use the dashboard's client (even though the result would be the same)
Also, we should check (`selfsubjectaccessreview`) if the user can get the openAPI before returning the (cached) result.

File: frontend/src/utils/api.js [hunk 3]
```diff
@@ -147,6 +147,10 @@ export function updateShootAddons ({ namespace, name, data }) {
   return updateResource(`/api/namespaces/${namespace}/shoots/${name}/spec/addons`, data)
 }
 
+export function getShootSpec () {
+  return getResource(`/api/info/shootspec`)
```
Comment (petersutter) [inline frontend/src/utils/api.js]:
```suggestion
  return getResource('/api/info/shootspec')
```

File: frontend/src/components/ShootEditor.vue [hunk 4]
```diff
@@ -290,6 +318,35 @@ export default {
...
+          }
+          const tooltip = this.shootEditorCompletions.editorTooltip(e, cm)
+          if (tooltip) {
+            this.helpTooltip.visible = true
+            this.helpTooltip.posX = e.clientX
+            this.helpTooltip.posY = e.clientY
+            this.helpTooltip.property = tooltip.property
+            this.helpTooltip.type = tooltip.type
+            this.helpTooltip.description = tooltip.description
+          }
```
Comment (petersutter) [inline frontend/src/components/ShootEditor.vue]:
```suggestion
          if (!tooltip) {
            return
          }
          this.helpTooltip.visible = true
          this.helpTooltip.posX = e.clientX
          this.helpTooltip.posY = e.clientY
          this.helpTooltip.property = tooltip.property
          this.helpTooltip.type = tooltip.type
          this.helpTooltip.description = tooltip.description
```

File: frontend/src/components/ShootEditor.vue [hunk 5]
```diff
@@ -290,6 +318,35 @@ export default {
...
+
+      CodeMirror.registerHelper('hint', 'yaml', (editor, options) => {
+        options.completeSingle = false
+        if (!this.shootEditorCompletions) {
+          return
+        }
+        return this.shootEditorCompletions.yamlHint(editor)
+      })
+
+      let cmTooltipFnTimerID
```
Comment (petersutter) [inline frontend/src/components/ShootEditor.vue]: shorten to timeoutId

File: frontend/src/components/ShootEditor.vue [hunk 6]
```diff
@@ -290,6 +318,35 @@ export default {
...
+      const cm = this.$instance
+      CodeMirror.on(element, 'mouseover', (e) => {
+        clearTimeout(cmTooltipFnTimerID)
+        this.helpTooltip.visible = false
+        cmTooltipFnTimerID = setTimeout(() => {
```
Comment (petersutter) [inline frontend/src/components/ShootEditor.vue]: set a timeout

File: frontend/src/utils/api.js [hunk 7]
```diff
@@ -147,6 +147,10 @@ export function updateShootAddons ({ namespace, name, data }) {
   return updateResource(`/api/namespaces/${namespace}/shoots/${name}/spec/addons`, data)
 }
 
+export function getShootSpec () {
```
Comment (petersutter) [inline frontend/src/utils/api.js]: maybe we find a better name?

File: frontend/src/components/ShootEditor.vue [hunk 8]
```diff
@@ -348,10 +405,14 @@ export default {
...
     }
   },
-  mounted () {
+  async mounted () {
     this.createInstance(this.$refs.container)
     this.update(this.value)
     this.refresh()
+
+    let shootCompletions = await getShootSpec()
+    shootCompletions = get(shootCompletions, 'data.spec.properties', {})
```
Comment (petersutter) [inline frontend/src/components/ShootEditor.vue]: better name than shootCompletions? This is the openApi spec for a shoot, right?

File: backend/lib/routes/info.js (hunk 2 above)
Comment (grolu) [reply backend/lib/routes/info.js]: The question is if and how we want to cache the result. But you are right, we should not return any information fetched with a user token to any other user. I will have a look at the issue

Comment (grolu) [reply backend/lib/routes/info.js]: Changed it to use the dashboardClient (admin) to fetch the Open API Spec. However, it is not possible to do a selfsubjectaccessreview as it is not a k8s resource (similar to healthz endpoint).
The only requirement is that the user needs to be authorized...

Comment (grolu) [reply backend/lib/routes/info.js]: Ok I added a access review for the non resource endpoint

File: frontend/src/utils/api.js (hunk 7 above)
Comment (grolu) [reply frontend/src/utils/api.js]: renamed

File: frontend/src/components/ShootEditor.vue (hunk 8 above)
Comment (grolu) [reply frontend/src/components/ShootEditor.vue]: renamed

File: backend/lib/kubernetes-client/nonResourceEndpoints/index.js [hunk 9]
```diff
@@ -17,10 +17,12 @@
 'use strict'
 
 const Healthz = require('./Healthz')
+const OpenAPI = require('./OpenAPI')
 
 function load (options) {
   return {
-    healthz: new Healthz(options)
+    healthz: new Healthz(options),
+    openAPI: new OpenAPI(options)
```
Comment (holgerkoser) [inline backend/lib/kubernetes-client/nonResourceEndpoints/index.js]: route should be `openapi` not `openAPI`

File: backend/lib/services/authorization.js [hunk 10]
```diff
@@ -38,21 +39,33 @@ async function hasAuthorization (user, resourceAttributes) {
...
-    resource: 'projects'
+    resourceAttributes: {
+      verb: 'create',
+      group: 'core.gardener.cloud',
+      resource: 'projects'
+    }
+  })
+}
+
+exports.canAccessOpenAPI = function (user) {
```
Comment (holgerkoser) [inline backend/lib/services/authorization.js]: rename canGetOpenAPI

File: backend/lib/routes/info.js [hunk 11]
```diff
@@ -58,3 +61,33 @@ async function fetchGardenerVersion () {
...
+  if (!await canAccessOpenAPI(user)) {
+    throw new Forbidden('User is not allowed to fetch the Open API specification')
+  }
+
+  if (shootOpenAPISpecification) {
+    return shootOpenAPISpecification
+  }
+
+  // Do not use client of user as the result gets cached and returned to other users
+  const res = await dashboardClient.openAPI.get()
```
Comment (holgerkoser) [inline backend/lib/routes/info.js]: we always use lowercase resource and nonresources names

File: backend/lib/routes/info.js [hunk 12]
```diff
@@ -22,8 +22,11 @@ const logger = require('../logger')
 const { decodeBase64 } = require('../utils')
 const { dashboardClient, isHttpError } = require('../kubernetes-client')
 const { version } = require('../../package')
-
+const { canAccessOpenAPI } = require('../services/authorization')
```
Comment (holgerkoser) [inline backend/lib/routes/info.js]: move openapi shoot spec to own file/folder next to healthz.

File: backend/lib/routes/info.js (hunk 11 above)
Comment (holgerkoser) [reply backend/lib/routes/info.js]: rename to `shootOpenApiSchema` or `shootSchema` and `getShootSchema`.

File: backend/lib/routes/info.js [hunk 13]
```diff
@@ -58,3 +61,33 @@ async function fetchGardenerVersion () {
...
+
+  if (shootOpenAPISpecification) {
+    return shootOpenAPISpecification
+  }
+
+  // Do not use client of user as the result gets cached and returned to other users
+  const res = await dashboardClient.openAPI.get()
+  const spec = await SwaggerParser.dereference(res)
+
+  shootOpenAPISpecification = _.get(spec, ['definitions', 'com.github.gardener.gardener.pkg.apis.core.v1alpha1.Shoot'], {})
```
Comment (holgerkoser) [inline backend/lib/routes/info.js]: v1alpha1 or v1beta1?

File: backend/lib/routes/info.js [hunk 14]
```diff
@@ -58,3 +61,33 @@ async function fetchGardenerVersion () {
     throw err
   }
 }
+
+router.route('/openapi/shoot')
```
Comment (holgerkoser) [inline backend/lib/routes/info.js]: move one level up.

Comment (holgerkoser) [reply backend/lib/routes/info.js]: I propose to have a route a route `openapi` or `schemas`  which returns all schemas required by the frontend. In our case:
```json
{ 
  "definitions": {
   "com.github.gardener.gardener.pkg.apis.core.v1alpha1.Shoot": {
      ...
    }
  }
}
```

File: frontend/src/utils/api.js (hunk 7 above)
Comment (holgerkoser) [reply frontend/src/utils/api.js]: getShootSchema

File: frontend/src/utils/shootEditorCompletions.js [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+    return {
+      list,
+      from: CodeMirror.Pos(cur.line, token.start),
+      to: CodeMirror.Pos(cur.line, token.end)
+    }
+  }
+
+  // Callback function for CodeMirror tooltip
+  editorTooltip (e, cm) {
+    let pos = cm.coordsChar({
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: const

File: frontend/src/utils/shootEditorCompletions.js [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+      }
+    }
+  }
+
+  // callback function for cm editor enter function
+  editorEnter (cm) {
+    const cur = cm.getCursor()
+
+    const { lineString, lineTokens } = this._getTokenLine(cm, cur)
+    const [, indent, firstArrayItem] = lineString.match(/^(\s*)(-\s)?(.*?)?$/)
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
    const [, indent, firstArrayItem] = lineString.match(/^(\s*)(-\s)?(.*?)?$/) || []
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    const completionPath = this._getTokenCompletionPath(token, cur, cm)
+
+    let completions
+    if (completionPath.length > 0) {
+      completions = get(this.shootCompletions, completionPath)
+    } else {
+      completions = this.shootCompletions
+    }
+
+    let completionArray = []
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: const

File: frontend/src/utils/shootEditorCompletions.js [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    const generateCompletionText = (propertyName, yamlType, tokenType) => {
+      const completionIndentStr = `${repeat(' ', token.start)}${repeat(' ', this.indentUnit)}${repeat(' ', tokenType === 'firstArrayItem' ? this.arrayBulletIndent : 0)}`
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: to much inline. extract values in single line.

File: frontend/src/utils/shootEditorCompletions.js [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+      if (yamlType === 'array') {
+        return `${propertyName}:\n${completionIndentStr}- `
+      } else if (yamlType === 'object') {
+        return `${propertyName}:\n${completionIndentStr}`
+      } else {
+        return `${propertyName}: `
+      }
+    }
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
      if (yamlType === 'array') {
        return `${propertyName}:\n${completionIndentStr}- `
      }  
      if (yamlType === 'object') {
        return `${propertyName}:\n${completionIndentStr}`
      }
      return `${propertyName}: `
    }
```

File: frontend/src/utils/shootEditorCompletions.js (hunk 17 above)
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
    const completionArray = []
```

File: frontend/src/utils/shootEditorCompletions.js (hunk 18 above)
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
      const numberOfSpaces = token.start + this.indentUnit + (tokenType === 'firstArrayItem' ? this.arrayBulletIndent : 0)
      const completionIndentStr = repeat(' ', numberOfSpaces)
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+
+    forIn(completions, (completion, propertyName) => {
+      const text = `${token.type === 'firstArrayItem' ? '- ' : ''}${generateCompletionText(propertyName, completion.type, token.type)}`
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
      let text = generateCompletionText(propertyName, completion.type, token.type)
      if (token.type === 'firstArrayItem') {
        text = '-' + text
      }
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+      const string = propertyName.toLowerCase()
+      completionArray.push({
+        text,
+        string,
+        property: propertyName,
+        type: upperFirst(completion.type),
+        description: completion.description,
+        render: (el, self, data) => {
+          const propertyWrapper = document.createElement('div')
+          propertyWrapper.innerHTML = `<span class="property">${propertyName}</span><span class="type">${upperFirst(completion.type)}</span>`
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: XSS?

File: frontend/src/utils/shootEditorCompletions.js [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+          propertyWrapper.className = 'ghint-type'
+          el.appendChild(propertyWrapper)
+
+          const descWrapper = document.createElement('div')
+          descWrapper.innerHTML = `<span class="description">${completion.description}</span>`
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: escape html!!!

File: frontend/src/utils/shootEditorCompletions.js [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+          tokenContext.unshift(currentToken)
+        }
+      }
+      line--
+    }
+
+    // token context contains all parrent tokens of `token, except for the token itself
+    const tokenPath = flatMap(tokenContext, (pathToken, index, tokenContext) => {
+      const isLeafContextToken = pathToken === last(tokenContext)
+      if (pathToken.type === 'property') {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: avoid deep nesting

File: frontend/src/utils/shootEditorCompletions.js [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+  _getTokenCompletionPath (token, cur, cm) {
+    let currentToken = token
+    let line = cur.line
+    let tokenContext = []
+
+    while (line >= 0 && !this._isTopLevelProperty(currentToken)) {
+      currentToken = this._getYamlToken(cm, CodeMirror.Pos(line, 0))
+      if (this._isCurrentTokenParentOfToken(currentToken, token) &&
+        this._isCurrentTokenIndentSmallerThanContextRoot(currentToken, tokenContext)) {
+        if (currentToken.type === 'property') {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: switch

File: frontend/src/utils/shootEditorCompletions.js [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+      token.start = token.end - token.string.length
+    }
+    return token
+  }
+
+  // returns path to completions in shootCompletions object
+  _getTokenCompletionPath (token, cur, cm) {
+    let currentToken = token
+    let line = cur.line
+    let tokenContext = []
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: const

File: frontend/src/utils/shootEditorCompletions.js [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+        const isTokenIndentIndicatingObjectStart = token.start === pathToken.indent + this.indentUnit + this.arrayBulletIndent
+        if (pathToken.propertyName !== undefined && isLeafContextToken && isTokenIndentIndicatingObjectStart) {
+          // firstArrayItem line can also start new object, so list properties of item object
+          return ['items', 'properties', pathToken.propertyName, 'properties']
+        }
+        // path token is array, so list properties of its items
+        return ['items', 'properties']
+      } else {
+        return []
+      }
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
  switch (pathToken.type) {
    case 'property': {
      const nextToken = nth(tokenContext, index + 1)
      if (nextToken && nextToken.type === 'firstArrayItem') {
        // next item is array, so don't append 'properties' to path
        return pathToken.propertyName
      }
      if (isLeafContextToken && token.type === 'firstArrayItem') {
        // leaf context token is array, so list properties of its items
        return [pathToken.propertyName, 'items', 'properties']
      }
      // regular property token
      return [pathToken.propertyName, 'properties']
    }
    case 'firstArrayItem': {
      const isTokenIndentIndicatingObjectStart = token.start === pathToken.indent + this.indentUnit + this.arrayBulletIndent
      if (pathToken.propertyName !== undefined && isLeafContextToken && isTokenIndentIndicatingObjectStart) {
        // firstArrayItem line can also start new object, so list properties of item object
        return ['items', 'properties', pathToken.propertyName, 'properties']
      }
      // path token is array, so list properties of its items
      return ['items', 'properties']
    }
  }
  return []
```

File: frontend/src/utils/shootEditorCompletions.js (hunk 25 above)
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
    const tokenContext = []
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 27]
```diff
@@ -0,0 +1,293 @@
...
+    // token context contains all parrent tokens of `token, except for the token itself
+    const tokenPath = flatMap(tokenContext, (pathToken, index, tokenContext) => {
+      const isLeafContextToken = pathToken === last(tokenContext)
+      if (pathToken.type === 'property') {
+        const nextToken = nth(tokenContext, index + 1)
+        if (nextToken && nextToken.type === 'firstArrayItem') {
+          // next item is array, so don't append 'properties' to path
+          return pathToken.propertyName
+        }
+        if (isLeafContextToken && token.type === 'firstArrayItem') {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]: token or nextToken or currentToken? I am not sure?

File: frontend/tests/unit/utils.shootEditorCompletions.spec.js [hunk 28]
```diff
@@ -0,0 +1,316 @@
...
+
+import repeat from 'lodash/repeat'
+
+import chai from 'chai'
+import sinon from 'sinon'
+import sinonChai from 'sinon-chai'
+chai.use(sinonChai)
+const expect = chai.expect
+
+document.body.createTextRange = function () {
```
Comment (holgerkoser) [inline frontend/tests/unit/utils.shootEditorCompletions.spec.js]: sandbox

File: backend/lib/routes/info.js (hunk 13 above)
Comment (grolu) [reply backend/lib/routes/info.js]: has been updated in the meantime, I will switch to v1beta1

File: frontend/src/utils/shootEditorCompletions.js (hunk 21 above)
Comment (grolu) [reply frontend/src/utils/shootEditorCompletions.js]: escaped using lodash escape

File: frontend/src/utils/shootEditorCompletions.js (hunk 22 above)
Comment (grolu) [reply frontend/src/utils/shootEditorCompletions.js]: escaped using lodash escape

File: frontend/src/utils/shootEditorCompletions.js (hunk 27 above)
Comment (grolu) [reply frontend/src/utils/shootEditorCompletions.js]: token is correct

Comment (grolu) [reply frontend/src/utils/shootEditorCompletions.js]: Do you want to go over this in more detail?

File: frontend/tests/unit/utils.shootEditorCompletions.spec.js (hunk 28 above)
Comment (grolu) [reply frontend/tests/unit/utils.shootEditorCompletions.spec.js]: cannot use sandbox as stubbing non existant functions is not possible in sinon. However, I can remove the function after the tests in an after() hook.

File: backend/lib/openapi/index.js [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+const _ = require('lodash')
+
+const router = module.exports = express.Router()
+
+router.route('/shoot')
+  .get(async (req, res, next) => {
+    try {
+      const user = req.user
+      const spec = await getShootSchema(user)
+      res.send({ spec })
```
Comment (holgerkoser) [inline backend/lib/openapi/index.js]:
```suggestion
      const schema = await getShootSchema(user)
      res.send(schema)
```

File: backend/lib/openapi/index.js [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+    return shootSchema
+  }
+
+  // Do not use client of user as the result gets cached and returned to other users
+  const res = await dashboardClient.openapi.get()
+  const spec = await SwaggerParser.dereference(res)
+
+  shootSchema = _.get(spec, ['definitions', 'com.github.gardener.gardener.pkg.apis.core.v1beta1.Shoot'], {})
+  return shootSchema
+}
```
Comment (holgerkoser) [inline backend/lib/openapi/index.js]:
```suggestion
router.route('/')
  .get(async (req, res, next) => {
    try {
      const user = req.user
      const schemaDefinitions = await getSchemaDefinitions(user)
      res.send(schemaDefinitions)
    } catch (err) {
      next(err)
    }
  })

const schemaDefinitions = {} // Cache, TODO: Need to update cache when apiserver gets updated

async function getSchemaDefinitions (user) {
  const hasAuthorization = await canGetOpenAPI(user)
  if (!hasAuthorization) {
    throw new Forbidden('User is not allowed to read OpenAPI schemas definitions')
  }

  if (_.isEmpty(schemaDefinitions)) {
    // Do not use client of user as the result gets cached and returned to other users
    const swaggerApi = await dashboardClient.openapi.get()
    const dereferencedSwaggerApi  = await SwaggerParser.dereference(swaggerApi)

    const selectedSchemaDefinitions = _
      .chain(dereferencedSwaggerApi)
      .get('definitions')
      .pick([
        'com.github.gardener.gardener.pkg.apis.core.v1beta1.Shoot'
      ])
      .value()
    _.assign(schemaDefinitions, selectedSchemaDefinitions)
  }

  return schemaDefinitions
}
```

Comment (holgerkoser) [reply backend/lib/openapi/index.js]: Return all schema definitions to the ui. Currently only shoots. Maybe more in the future. May we should map to apigroup??

File: frontend/src/utils/api.js [hunk 31]
```diff
@@ -148,7 +148,7 @@ export function updateShootAddons ({ namespace, name, data }) {
 }
 
 export function getShootSpec () {
-  return getResource('/api/info/openapi/shoot')
+  return getResource('/api/openapi/shoot')
```
Comment (holgerkoser) [inline frontend/src/utils/api.js]:
```suggestion
export function getShootSchemaDefinition () {
  const definitions = await getResource('/api/openapi')
  return definitions['com.github.gardener.gardener.pkg.apis.core.v1beta1.Shoot']
}
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 32]
```diff
@@ -134,12 +139,15 @@ export class ShootEditorCompletions {
         description: completion.description,
         render: (el, self, data) => {
           const propertyWrapper = document.createElement('div')
-          propertyWrapper.innerHTML = `<span class="property">${propertyName}</span><span class="type">${upperFirst(completion.type)}</span>`
+          const escapedType = escape(upperFirst(completion.type))
+          const escapedPropertyName = escape(propertyName)
+          const escapedDescription = escape(completion.description)
+          propertyWrapper.innerHTML = `<span class="property">${escapedPropertyName}</span><span class="type">${escapedType}</span>`
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
          propertyWrapper.innerHTML = [ 
            '<span class="property">',
            escape(propertyName)
            '</span>',
            '<span class="type">',
            escape(upperFirst(completion.type))
            '</span>'
          ].join('')
```

Comment (holgerkoser) [reply frontend/src/utils/shootEditorCompletions.js]: I think readability is better

File: frontend/tests/unit/utils.shootEditorCompletions.spec.js [hunk 33]
```diff
@@ -106,10 +71,59 @@ const shootCompletions = {
...
   describe('shootEditorCompletions', function () {
+    let editor
+    let setEditorContentAndCursor
+    let setEditorCursor
+    let shootEditorCompletions
+
+    before(function () {
+      // cannot stub via sandbo as Jest does not implement the function and
+      // non existant functions cannot be stubbed using a sinon sandbox
+      document.body.createTextRange = function () {
```
Comment (holgerkoser) [inline frontend/tests/unit/utils.shootEditorCompletions.spec.js]: stub with sandbox

Comment (grolu) [reply frontend/tests/unit/utils.shootEditorCompletions.spec.js]: not possible, see https://github.com/sinonjs/sinon/pull/1557
Or how would you do this?

File: frontend/src/components/ShootEditor.vue [hunk 34]
```diff
@@ -1,4 +1,4 @@
-<!--
+s<!--
```
Comment (petersutter) [inline frontend/src/components/ShootEditor.vue]:
```suggestion
<!--
```

File: backend/test/acceptance/openapi.spec.js [hunk 35]
```diff
@@ -30,16 +30,17 @@ module.exports = function ({ agent, k8s, auth }) {
...
     const res = await agent
-      .get('/api/openapi/shoot')
+      .get('/api/openapi')
       .set('cookie', await user.cookie)
 
     expect(res).to.have.status(200)
     expect(res).to.be.json
-    expect(res.body).to.have.property('spec').that.is.eql({ spec: { type: 'object' } })
+    const shootDefinition = res.body['com.github.gardener.gardener.pkg.apis.core.v1beta1.Shoot']
+    expect(shootDefinition).to.have.property('spec').that.is.eql({ type: 'object' })
```
Comment (holgerkoser) [inline backend/test/acceptance/openapi.spec.js]: stimmt nicht oder?

File: frontend/src/components/ShootEditor.vue [hunk 36]
```diff
@@ -422,8 +422,8 @@ export default {
     this.update(this.value)
     this.refresh()
 
-    const shootOpenAPISpecification = await getShootSpec()
-    const shootProperties = get(shootOpenAPISpecification, 'data.spec.properties', {})
+    const shootOpenAPISpecification = await getShootSchemaDefinition()
+    const shootProperties = get(shootOpenAPISpecification, 'properties', {})
```
Comment (holgerkoser) [inline frontend/src/components/ShootEditor.vue]:
```suggestion
    const shootSchemaDefinition = await getShootSchemaDefinition()
    const shootProperties = get(shootSchemaDefinition, 'properties', {})
```

File: backend/test/acceptance/openapi.spec.js (hunk 35 above)
Comment (grolu) [reply backend/test/acceptance/openapi.spec.js]: naja typischer weise würde da properties: {}  kommen.... kann es ändern wenn ihr wollt

Comment (holgerkoser) [reply backend/test/acceptance/openapi.spec.js]: It does not have a property named `spec` anymore.

**Comment by grolu** [comment]
> Ok I will look into this

@petersutter can you give feedback if the latest changes resolve your issue?

**Comment by petersutter** [comment]
yes it's solved

File: frontend/src/utils/shootEditorCompletions.js [hunk 37]
```diff
@@ -0,0 +1,316 @@
...
+        if (currentToken.type === 'property') {
+          tokenContext.unshift(currentToken)
+        } else if (currentToken.type === 'firstArrayItem') {
+          tokenContext.unshift(currentToken)
+        }
+      }
+      line--
+    }
+
+    // token context contains all parrent tokens of `token, except for the token itself
```
Comment (petersutter) [inline frontend/src/utils/shootEditorCompletions.js]:
```suggestion
    // token context contains all parent tokens of `token, except for the token itself
```

Comment (petersutter) [reply frontend/src/utils/shootEditorCompletions.js]: if this is fixed we can merge this PR

Comment (grolu) [reply frontend/src/utils/shootEditorCompletions.js]: done

**Pull request merged**